docs/source/explanations/optimization/internal_optimizers.rst

```diff
@@ -0,0 +1,208 @@
+- You get a parallelized and customizable numerical gradient if the user did not provide
+  a closed form gradient.
+- You can compare your optimizer with all the other estimagic optimizers by changing
+  only one line of code.
+
+All of this functionality is achieved by transforming a more complicated user provided
+problem into a simpler problem and then calling "internal optimizers" to solve the
+transformed problem.
+
+
+
+The Internal Optimizer Interface
+================================
+
+An internal optimizer is a function that minimizes an objective function it has two
+mandatory arguments:
+
+1. ``criterion_and_derivative`` (described below)
+2. ``x``: One dimensional numpy array with start parameters.
+
+Followed by two strongly encouraged but not absolutely mandatory arguments:
+
+1. lower_bounds: 1d numpy array with the same length as x with lower
+    bounds for the parameters. Can be -np.inf for unbound parameters.
+2. upper_bounds: 1d numpy array with the same length as x with upper
+    bounds for the parameters. Can be -np.inf for unbound parameters.
+
+Moreover, it can accept any number of additional arguments that are used by the
+optimizer. Neither the mandatory arguments, nor the bounds should have a default value.
+
+The only non-trivial argument is ``criterion_and_derivative``. This is a callable that
+accepts three arguments and returns the output of the user provided criterion function
+(float or dict), the output of a user provided or numerical derivative (np.ndarray) or
+both. For more details on valid outputs of the criterion function see
+:ref:`maximize_and_minimize`.
+
+The arguments  of criterion_and_derivative are:
+
+- x (np.ndarray): A one dimensional vector with free parameters
+- task (str): One of "criterion", "derivative" and "criterion_and_derivative"
+- algorithm_info (dict): Dict with the following entries:
+    - "primary_criterion_entry": For optimizers that minimize a scalar function this has
+      to be "value". For optimizers that optimize a sum it has to be "contributions".
+      For optimizers that optimize a sum of squares it has to be "root_contributions".
+    - "parallelizes": Bool that indicates if the algorithm calls the internal
+      criterion function in parallel. If so, caching is disabled.
+    - "needs_scaling": Bool. True if the algorithm is not scale invariant. In that case
+      we can issue a warning if the user did not scale his problem with estimagic.
+    - "name": The name of the algorithm that can be displayed in error messages
+
+``criterion_and_derivative`` is the centerpiece of how estimagic achieves its magic.
+It does logging, converts the internal parameters into a DataFrame for the user provided
+criterion function, handles errors and more.
+
+.. _internal_optimizer_output:
+
+Output of Internal Optimizers
+=============================
+
+
+After convergence or when another stopping criterion is achieved the internal optimizer
```

Comment:
- What if a key cannot be provided?
- What is solution_criterion? Scalar, array?

Reply:
Clarified.